README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# RGremlinsConjoint

<!-- badges: start -->

[![R-CMD-check](https://github.com/statuser/RGremlinsConjoint/workflows/R-CMD-check/badge.svg)](https://github.com/statuser/RGremlinsConjoint/actions)
[![R-CMD-check](https://github.com/statuser/RGremlinsConjoint/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/statuser/RGremlinsConjoint/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

The gremlins package provides the tools and utilities to estimate a the
model described in “Gremlins in the Data: Identifying the Information
Content of Research Subjects”
(\[<https://doi.org/10.1177/0022243720965930>\]) using conjoint analysis
data such as that collected in Sawtooth Software’s Lighthouse or
Discover Products. The packages also contains utility functions for
formatting the input data and extracting the relevant results.

## Installation

<!-- You can install the released version of RGremlinsConjoint from [CRAN](https://CRAN.R-project.org) with: -->
<!-- ``` r -->
<!-- install.packages("RGremlinsConjoint") -->
<!-- ``` -->

You can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("statuser/RGremlinsConjoint")
```

## Example

The package exposes basically one function You can use it like:

``` r
library(RGremlinsConjoint)

# Read in the data
truck_design_file <- system.file("extdata", "simTruckDesign.csv", package = "RGremlinsConjoint")
truck_data_file <- system.file("extdata", "simTruckData.csv", package = "RGremlinsConjoint")
truckDesign <- read.csv(truck_design_file)
truckData <- read.csv(truck_data_file)

# Covert the design file to be dummy coded is necessary
# The simulated data is already coded
# codedTruck <- code_sawtooth_design(truckDesign, c(4:9), include_none_option=TRUE)

outputSimData_burn <- estimateGremlinsModel(truckData,
                                            truckDesign,
                                            R = 10,
                                            keepEvery = 1,
                                            num_lambda_segments = 2)
#> Finding Starting Values
#> Beginning MCMC Routine
#> Completing iteration :  1 
#> Accept rate slopes:  0 
#> Accept rate lambda:  0 
#> Mu_adapt lambda:     50 
#> Gamma_adapt lambda:  10 
#> metstd lambda:       10 
#> current lambda:      50
```


NEWS.md
# RGremlinsConjoint (development version)

# RGremlinsConjoint 0.9.1

* Fixed Documentation issue that prevented the package helpfile from generating


# RGremlinsConjoint 0.9.0

* Initial Release


tests/testthat.R
library(testthat)
library(RGremlinsConjoint)

test_check("RGremlinsConjoint")


R/RGremlinsConjoint.R
# INTRODUCTION
# The files in this directory constitute the R package behind the 'Gremlins in the Data'
# JMR paper. The scripts use MCMC to estimate parameters of the model proposed in the paper.
# Authors:
#   John R. Howell (jrhowell@byu.edu)
#   Peter Ebbes (ebbes@hec.fr)
#   John C. Liechty (jcl12@psu.edu)


#' 'RGremlinsConjoint': A package for estimating the "Gremlins in the Data" model
#'
#' The tools and utilities to estimate the model described in "Gremlin's in the
#' Data: Identifying the Information Content of Research Subjects" (Howell et al.
#' (2021) <doi:10.1177/0022243720965930>)  using conjoint analysis data such as
#' that collected in Sawtooth Software's 'Lighthouse' or 'Discover' products.
#' Additional utilities are included for formatting the input data.
#'
#' @docType package
#' @name gremlins
NULL


#' Set global options for the gremlins models.  These options are not expected to be modified by the user
#' but are extracted from the functions to simplify the coding.
gremlinsEnv <- new.env()

# TODO create functions to set these values
gremlinsEnv$jumpSizes <- c(0.05, 0.1, 0.5)
gremlinsEnv$jumpSizeProbs <- c(0.65, 0.25, 0.10)
gremlinsEnv$totalConstraintTries <- 100
gremlinsEnv$num_lambda_segments <- 2


#' Simulated data for the "Gremlins in the Data Model"
#'
#' A dataset containing simulated choices from a CBC study where some of the
#' respondents are information poor or 'Gremlins'.  The data is simulated data
#' and does not reflect actual preferences.
#'
#' @format A data frame with 32000 rows and 10 variables:
#' \describe{
#'   \item{resp.id}{A respondent identifier}
#'   \item{ques}{The question or task number}
#'   \item{alt}{The choice alternative}
#'   \item{choice}{An indicator that takes on a value of 1 if the alternative was chosen.  (Default is 0.)}
#'   \item{brandFord}{A dummy coded variable indicating the brand is Ford}
#'   \item{brandGM}{A dummy coded variable indicating the brand is GM}
#'   \item{brandDodge}{A dummy coded variable indicating the brand is Dodge}
#'   \item{enghyb}{A dummy coded variable indicating the engine is a hybrid}
#'   \item{engelec}{A dummy coded variable indicating the engine is electric}
#'   \item{price}{A continuous variable for the relative price of the individual offerings.}
#' }
#'
"cbc.df"


data-raw/cameraDataSetup.R
# Conjoint response: dual response
# Des of attributes: backward
# Number: "just right"
# Price Range: 79-279

data <- read.csv(file = "data-raw/RawCameraData.csv", header = TRUE, stringsAsFactors = FALSE)

# For respondents that picked the none option, there are two tasks for that
# choice. The first is X.0 and the second is X.1.  Where X is the scenario
# number.  The X.0 task will always be a none option. The X.1 task with be the
# preference choice not accounting for the none.  These two choices have a
# correlated error term and should not be analyzed as an extra piece of
# information. We avoid this problem by dropping the preference data and
# analyzing it as if the only choice made as the none option.
dataSingleChoice <- subset(data, !grepl("\\.1$", data$scenario, perl=TRUE))

# write.csv(dataSingleChoice, "CameraData.csv")


# Filter the survey times to just include the relevant data
# responseTime <- read.csv(file = "cellXTime.csv", header = TRUE, stringsAsFactors = FALSE)
# cameraResponseTimes <- subset(responseTime, responseTime$ID %in% unique(data$xsurvnum))
# dim(cameraResponseTimes)
# length(unique(data$xsurvnum))
# write.csv(cameraResponseTimes, "cameraTime.csv")


#Re-frame the data into Sawtooth Style Format.

nTotalTasks <- nrow(dataSingleChoice)/5 #5 choices per set including the none.
concepts <- rep(1:5, nTotalTasks)

library(tidyverse)
dataSingleChoice <- as_tibble(dataSingleChoice)
dataSingleChoice <- dataSingleChoice %>% add_column(concept = concepts, .before = "y1")

cameraResponses <- dataSingleChoice %>%
  select(xsurvnum, version, scenario, concept, y1) %>%
  mutate(scenario, scenario = paste("Task", scenario, sep ="")) %>%
  filter(y1 == 1) %>%
  select(-y1) %>%
  spread(scenario, concept) %>%
  select(xsurvnum, version, Task1, Task2, Task3, Task4, Task5, Task6, Task7,
         Task8, Task9, Task10, Task11, Task12, Task13, Task14, Task15, Task16)

cameraDesign <- dataSingleChoice %>%
  select(-y1) %>%
  mutate(brand = brand_1 + 2*brand_2 + 3*brand_3 + 4*brand_4) %>%
  select(-brand_1, -brand_2, -brand_3, -brand_4) %>%
  mutate(pixels = pixels_2 + 1, zoom = zoom_2 + 1, video = video_2 + 1, swivel=swivel_2 + 1, wifi=wifi_2 + 1) %>%
  select(-pixels_2, -zoom_2, -video_2, -swivel_2, -wifi_2) %>%
  filter(concept != 5) %>%
  select(version, scenario, concept, brand, pixels, zoom, video, swivel, wifi, price_lin) %>%
  mutate(price_lin = round(exp(price_lin)/100, digits=2)) %>%
  filter(!duplicated(.)) %>%
  arrange(version, scenario, concept)

# Convert Prices to be factor levels
cameraDesign <- cameraDesign %>%
  mutate(price_lin = as_factor(price_lin)) %>%
  mutate(price_lin = fct_recode(price_lin, "1" = "0.79", "2" = "1.29", "3" = "1.79", "4" = "2.29", "5" = "2.79")) %>%
  mutate(price_lin = as.numeric(price_lin))



write_csv(cameraDesign, "data-raw/CameraDesign.csv")

write_csv(cameraResponses, "data-raw/CameraFullData.csv")


data-raw/Example Analysis.R
# To install the package you will need to run the following command
# devtools::install_github("statuser/RGremlinsConjoint")
library(RGremlinsConjoint)

truck_design_file <- system.file("extdata", "simTruckDesign.csv", package = "RGremlinsConjoint")
truck_data_file <- system.file("extdata", "simTruckData.csv", package = "RGremlinsConjoint")

# Read in the Sawtooth Formatted data
truckDesign <- read.csv(truck_design_file)
truckData <- read.csv(truck_data_file)

## Covert the design file to be dummy coded
# the truck data is already coded
## codedTruck <- code_sawtooth_design(truckDesign, c(4:9), include_none_option=TRUE)


# Tune Atchade algorithm; monitor accept rates; doesn't update if total MCMC < 1000
# Adjust Atch_tau_tune_slopes and Atch_tau_tune_lambda until accept rates are about 20-80%
# Put those numbers in the function input of the burn in


outputSimData_burn <- estimateGremlinsModel(truckData,
                                            truckDesign,
                                            R = 1010,
                                            keepEvery = 1,
                                            num_lambda_segments = 2,
                                            Atchade_lambda_tuning = 1)



R/estimateGremlinsModelHelperFunctions.R
#' Convert a Sawtooth Software generated design file to a dummy coded design
#' file
#'
#' Take a design file such as those generated by the Sawtooth Software
#' 'Lighthouse Studio' and convert it into a dummy coded design file.  The last
#' level in the attribute is considered the reference level and will be dropped.
#'
#' This function is written to not require converting columns to be factors. All
#' variables should be numeric indexes for the levels of the attributes. If you
#' would like to manually code a attribute of the design, for example if you
#' have a price variable, you must manually code that attribute and then can
#' call the function with the optional columns to code parameter.
#'
#' @seealso
#' /url{http://www.sawtoothsoftware.com/help/lighthouse-studio/manual/index.html?hid_web_cbc_exporting.html}
#'
#'
#' Documentation for the Sawtooth Software Design file format can be found at
#' @param sawtooth_design A matrix that contains the Sawtooth design.  Can be
#'   loaded with read.csv.
#' @param columns_to_code (Optional, Default = all columns) A vector listing the
#'   numeric index of the columns to code. Note: The first column is column 4
#'   due to the control variables
#' @param include_none_option (Optional, Default = FALSE) A boolean value
#'   indicating whether to expand the task to include a none option
#' @return A matrix object chat contains the dummy coded design file.  The last
#'   attribute is considered the reference level
#'
#' @export
#' @examples
#' \dontrun{
#'  # Read in the Sawtooth Formatted data
#'  design <- read.csv("Design.csv")
#'  prices = c(0.79, 1.29, 1.79, 2.29, 2.79)
#'  design$price <- prices[design$price]
#'  codedDesign <- codeSawtoothDesignFile(design, c(4:9))
#' }
#'
#'
code_sawtooth_design <- function(sawtooth_design, columns_to_code = c(4:ncol(sawtooth_design)), include_none_option=FALSE) {
  dummyCode <- function(column) {
    nLevels <- max(column)
    dummy_code <- diag(nLevels)
    dummy_code <- dummy_code[,-nLevels, drop=FALSE]
    result_columns <- dummy_code[column[[1]],, drop=FALSE]
    colnames(result_columns) <- paste0(names(column), 1:(nLevels - 1))
    result_columns
  }

  dummy_coded_columns = list()
  for(i in 1:ncol(sawtooth_design)) {
    if(i %in% columns_to_code ) {
      dummy_coded_columns[[i]] <- dummyCode(sawtooth_design[i])
    } else {
      dummy_coded_columns[[i]] <- sawtooth_design[i]
    }
  }

  coded_design <- do.call("cbind", dummy_coded_columns)
  if(include_none_option) {
    nVersions <- max(coded_design[,1])
    nScenarios <- max(coded_design[,2])
    nConcepts <- max(coded_design[,3])
    versions <- rep(1:nVersions, each=nScenarios)
    scenarios <- rep(1:nScenarios, times=nVersions)
    none_design <- cbind(versions, scenarios, nConcepts+1, matrix(0, ncol=ncol(coded_design) - 3, nrow =length(versions)))
    colnames(none_design) <- colnames(coded_design)
    coded_design <- rbind(coded_design, none_design)
    coded_design <- coded_design[order(coded_design[,1], coded_design[,2], coded_design[,3]),]
  }

  return(coded_design)
}

#' Convert 'RGremlinsConjoint' formatted Data to 'bayesm' format
#'
#' Converts a data file and a coded design file from the format expected by the
#' package to a format appropriate for estimation in 'bayesm'
#' \code{\link[bayesm]{rhierMnlRwMixture}}
#'
#' @param data The data.frame or matrix that contains the respondents answers
#' @param design The data.frame or matrix that contains the coded design
#'
#' @return lgtdata The list data structure for use with 'bayesm'
#'
#' @seealso \code{\link{code_sawtooth_design}}
#' @export
#' @examples
#' \dontrun{
#'  data <- read.csv("data.csv")
#'  design <- read.csv("design.csv")
#'  design <- code_sawtooth_design(design)
#'  convert_to_bayesm(data, design)
#' }
convert_to_bayesm <- function(data, design) {
  # lgtdata - a nResp list with
  # lgtdata[[i]]$y - a vector of respondent choices (1,..., p) of length nTasks
  # lgtdata[[i]]$x - a matrix design matrix for the ith unit
  nResp <- nrow(data)
  nTasks <- max(design[,2])
  nConcepts <- max(design[,3])

  lgtdata <- list()
  for(i in seq_len(nResp)) {
    versionNumber <- data[i,2]
    y <- c(unlist(data[i,-c(1,2)]))
    X <- subset(design, design[,1] == versionNumber)

    X <- as.matrix(X[,-c(1:3)])
    lgtdata[[i]] <- list(y = y, X = X)
  }

  list(lgtdata = lgtdata, p = nConcepts)
}


R/estimateGremlins.R
#' Estimate Gremlin's Model - Hierarchical MNL
#'
#' The function estimates the model described in "Gremlin's in the Data: Identifying the
#' Information Content of Research Subjects" (Howell et al.
#' (2021) <doi:10.1177/0022243720965930>) using a hierarchical multinomial logit model
#'
#'
#' @param data A matrix containing the raw data.
#'   The first column the respondent identifier, followed by the design number, the remaining
#'   columns indicate the choices for the tasks that coincide to the design file.
#' @param design A matrix representing the coded (dummy of effects) design file.  The design file should be
#'   formatted as a matrix with number of versions X number of tasks X number of alternatives
#'   rows and number of parameters + 3 columns.  The first column contains the version number,
#'   the second columns contains the task number, the third column contains the alternative,
#'   and the remaining columns contain the coded design.  A generic Sawtooth Software design
#'   file can be converted to this format using the \code{\link{code_sawtooth_design}} function.
#' @param Priors A data structure that contains the priors for to the model.  Can be null indicating
#'   the use of default priors or must contain a full prior specification.
#' @param R The number of repetitions in the chain
#' @param keepEvery saves every keepEvery-th draw for output
#' @param verbose Print intermediate results to the screen (default = TRUE)
#' @param num_lambda_segments (Default = 2) The number of segments for the scale factor
#' @param constraints (Optional) a vector of length n-param specifying the constraints
#'    to impose on the parameters or NULL.  a 1 indicates the parameter is constrained to be positive
#'    a -1 constrains to be  negative, and a 0 indicates no constraint.
#' @param startingValues (Optional) starting values to use for the MCMC algorithm.  This is a list of
#'    containing: slope = a nRespondent by nParamter matrix of slopes for the respondent
#'                slopeBar = a nParameter vector of the slopeBar parameter
#'                slopeCov = a nParameter by nParameter matrix containing the variance covariance matrix
#'                           for the slopeBar parameter
#'                lambda = a nSegment vector containing the starting values for the lambda parameters.  The first
#'                         element in the vector should be 1.
#'                segMembership = a nRespondent vector containing the segment membership for each respondent.
#'                phi_lambda = a nParameter vector containing the base probabilities that an individual belongs
#'                             to each segment.  Should sum to 1.
#' @param previous_iterations The number of previous iterations run.  This parameter is used for the Atchade
#'      adaptive MCMC step size algorithm.  This is used since the Atchade update does not happen for less
#'      than 1000 iterations.  (Default = 0)
#' @param Atchade_slope_tuning The tuning factor to use for Atchade step for the slopes parameter.  Larger values
#'      decrease the acceptance rate.  (Default = 0.01)
#' @param Atchade_lambda_tuning The tuning factor to use for the Atchade step for the lambda parameter.  Larger
#'      values decrease the acceptance rate. (Default = 10)
#' @return A data structure containing the draws from the complete MCMC chain
#'
#' @examples
#'
#' truck_design_file <- system.file("extdata", "simTruckDesign.csv", package = "RGremlinsConjoint")
#' truck_data_file <- system.file("extdata", "simTruckData.csv", package = "RGremlinsConjoint")
#'
#' truckDesign <- read.csv(truck_design_file)
#' truckData <- read.csv(truck_data_file)
#' outputSimData_burn <- estimateGremlinsModel(truckData,
#'                                             truckDesign,
#'                                             R = 10,
#'                                             keepEvery = 1,
#'                                             num_lambda_segments = 2)
#'
#' @export
#' @seealso \code{\link{code_sawtooth_design}}
estimateGremlinsModel <- function(data,
                                  design,
                                  Priors = NULL,
                                  R = NULL,
                                  keepEvery = 1,
                                  verbose = TRUE,
                                  num_lambda_segments = 2,
                                  constraints = NULL,
                                  startingValues = NULL,
                                  previous_iterations = 0,
                                  Atchade_slope_tuning = 0.1,
                                  Atchade_lambda_tuning = 10) {

  #Set global parameters used by many of the functions
  gremlinsEnv$num_parameters <- ncol(design) - 3
  gremlinsEnv$num_respondents <- nrow(data)
  gremlinsEnv$num_tasks <- max(design[,2])
  gremlinsEnv$num_concepts <- max(design[,3])
  gremlinsEnv$num_lambda_segments <- num_lambda_segments

  # Used in Atchade steps
  gremlinsEnv$eps1 <- 10^(-7)
  gremlinsEnv$eps2 <- 10^(-6)
  gremlinsEnv$A1 <- 10^7
  gremlinsEnv$tau_bar <- 0.28		    # Targeted acceptance rate MH for beta_i's
  gremlinsEnv$tau_bar_lambda <- 0.28  # Targeted acceptance rate MH for lambda's
  gremlinsEnv$acceptanceRate_lambda <- rep(0,times=num_lambda_segments)

  if (is.null(constraints)){
    gremlinsEnv$acceptanceRate_slopes <- matrix(0, nrow=nrow(data), ncol=1)       # Monitor accept rates for each type of constraint
  }else{
    gremlinsEnv$acceptanceRate_slopes <- matrix(0, nrow=nrow(data), ncol=length(unique(constraints)))       # Monitor accept rates for each type of constraint 0, -1, 1
  }



  # Ensure design and data are actually matrices instead of data frames
  design <- as.matrix(design)
  data <- as.matrix(data)

  #Set up covariates to just an intercept if not supplied
  covariates = as.matrix(rep(1, gremlinsEnv$num_respondents))
  gremlinsEnv$nCovariates = 1

  if(is.null(constraints)) {
    constraints = double(gremlinsEnv$num_parameters)
  } else {
    if(length(constraints) != gremlinsEnv$num_parameters || !all(constraints %in% c(-1, 0, 1))) {
      stop("constraints must contain a value fo reach parameter and
           can only be -1, 0, or 1 for negative, no constraint, or positive")
    }
  }

  #validate and Set priors
  if(is.null(Priors)) {
    Priors = list(
      mu_not = matrix(0, ncol = gremlinsEnv$num_parameters, nrow = gremlinsEnv$nCovariates),
      Ainv = solve(1000*diag(ncol(covariates))),

      nu_not = gremlinsEnv$num_parameters + 2,
      V_not = 1*diag(gremlinsEnv$num_parameters),

      lambdaScale = c(1, rep(5, gremlinsEnv$num_lambda_segments - 1)),
      lambdaShape = c(1, rep(5, gremlinsEnv$num_lambda_segments - 1)),

      psi_k = rep(1, gremlinsEnv$num_lambda_segments)

    )
  } else {
    validatePriors(Priors)
  }



  respDesign <- list()
  respData <- list()

  for(ind in 1:gremlinsEnv$num_respondents) {
    respData[[ind]] <- as.matrix(data[ind, -c(1:2)]) # Drop the RespID and Design Version
    respDesign[[ind]] <- as.matrix(design[design[,1] == data[ind, 2], -c(1:3)]) # Drop the Version, Task, and Concept Columns
  }

  if(is.null(startingValues)) {
    if(verbose) {
      cat("Finding Starting Values\n")
    }
    startingValues <- find_starting_values(respData, respDesign)
  } else {
    # TODO: Validate starting values
  }

  # Initialization Atchade parameters; for each respondent a list
  Atch_MCMC_list_slopes <- generate_starting_atchade_slopes(startingValues$slopeBar,0.1)


  # Initialization Atchade parameters for lambda
  Atch_MCMC_list_lambda <- generate_starting_atchade_lambdas(10, startingValues$lambda)

  # Initialize Storage and MCMC
  slope <- array(0, dim=c(R%/%keepEvery, gremlinsEnv$num_respondents, gremlinsEnv$num_parameters))
  slope_MCMC <- startingValues$slope

  slopeBar <- array(0, dim=c(R%/%keepEvery, gremlinsEnv$num_parameters))
  slopeBar_MCMC <- matrix(startingValues$slopeBar, nrow = 1)

  slopeCov <- array(0, dim=c(R%/%keepEvery, gremlinsEnv$num_parameters, gremlinsEnv$num_parameters))
  slopeCov_MCMC <- startingValues$slopeCov

  lambda <- matrix(0, nrow = R%/%keepEvery, ncol = gremlinsEnv$num_lambda_segments)
  lambda_MCMC <- startingValues$lambda

  K <- matrix(0, nrow = R%/%keepEvery, ncol = gremlinsEnv$num_respondents)
  K_MCMC <- startingValues$segMembership

  phi_lambda <- matrix(0, nrow = R%/%keepEvery, ncol = gremlinsEnv$num_lambda_segments)
  phi_lambda_MCMC <- startingValues$phi_lambda



  if(verbose) {
    cat("Beginning MCMC Routine\n")
  }

  # Run MCMC
  for(rep in 1:R) {

      #; Provide status update on code
      if(rep %% 1000 == 1 && verbose) {
        cat(paste("Completing iteration : ", rep, "\n"))
        cat(paste("Accept rate slopes: ", colMeans(gremlinsEnv$acceptanceRate_slopes/rep), "\n"))
        cat(paste("Accept rate lambda: ", gremlinsEnv$acceptanceRate_lambda[2:num_lambda_segments]/rep, "\n"))
        cat(paste("Mu_adapt lambda:    ", Atch_MCMC_list_lambda$mu_adapt[2:num_lambda_segments], "\n"))
        cat(paste("Gamma_adapt lambda: ", Atch_MCMC_list_lambda$Gamma_adapt[2:num_lambda_segments], "\n"))
        cat(paste("metstd lambda:      ", Atch_MCMC_list_lambda$metstd[2:num_lambda_segments], "\n"))
        cat(paste("current lambda:     ", lambda_MCMC[2:num_lambda_segments], "\n"))

      }

      for(ind in 1:gremlinsEnv$num_respondents) {

        slope_atch_list_rep_ind <- generateIndividualSlope_ATCH(respData[[ind]],
                                                                respDesign[[ind]],
                                                                slope_MCMC[ind, ],
                                                                lambda_MCMC[K_MCMC[ind]],
                                                                slopeBar_MCMC,
                                                                slopeCov_MCMC, constraints,
                                                                ind_in = ind,
                                                                cur_mcmc_iter = (previous_iterations+rep),
                                                                metstd_in = Atch_MCMC_list_slopes[[ind]]$metstd,
                                                                Vmet_in = Atch_MCMC_list_slopes[[ind]]$Vmet,
                                                                Gamma_adapt_in = Atch_MCMC_list_slopes[[ind]]$Gamma_adapt,
                                                                mu_adapt_in = Atch_MCMC_list_slopes[[ind]]$mu_adapt
        )

        slope_MCMC[ind, ] <- slope_atch_list_rep_ind$currentSlope_out
        Atch_MCMC_list_slopes[[ind]] <- slope_atch_list_rep_ind$Atch_out_respi_list


        K_MCMC[ind] <- generateSegmentMembership(respData[[ind]], respDesign[[ind]], slope_MCMC[ind, ], lambda_MCMC, phi_lambda_MCMC, Priors)
      }

      multiregOut <- multivariateRegression(slope_MCMC, as.matrix(covariates), Priors)
      slopeBar_MCMC <- multiregOut$Beta
      slopeCov_MCMC <- multiregOut$Sigma

      lambda_atch_list_rep <- generateLambdaMix_ATCH(lambda_MCMC, respData, respDesign, slope_MCMC, K_MCMC, Priors,
                                                     cur_mcmc_iter = (previous_iterations+rep),
                                                     metstd_in = Atch_MCMC_list_lambda$metstd,
                                                     Vmet_in = Atch_MCMC_list_lambda$Vmet,
                                                     Gamma_adapt_in = Atch_MCMC_list_lambda$Gamma_adapt,
                                                     mu_adapt_in = Atch_MCMC_list_lambda$mu_adapt)


      lambda_MCMC <- lambda_atch_list_rep$lambda_out
      Atch_MCMC_list_lambda <- lambda_atch_list_rep$Atch_out_lambda_list


      phi_lambda_MCMC <- drawPhi(K_MCMC, Priors$psi_k)





    # Store only those draws that we intend to keep
    if(rep %% keepEvery == 0) {
      savedIterNum = rep %/% keepEvery

      slope[savedIterNum,,] <- slope_MCMC
      slopeBar[savedIterNum,] <- slopeBar_MCMC
      slopeCov[savedIterNum,,] <- slopeCov_MCMC
      lambda[savedIterNum,] <- lambda_MCMC
      K[savedIterNum,] <- K_MCMC
      phi_lambda[savedIterNum, ] <- phi_lambda_MCMC
    }


  }


  #; Save the output

  return(list(slope = slope,
              slopeBar = slopeBar,
              slopeCov = slopeCov,
              lambda = lambda,
              segmentMembership = K,
              lambdaSegProb = phi_lambda,
              Atch_MCMC_list_slopes = Atch_MCMC_list_slopes,
              Atch_MCMC_list_lambda = Atch_MCMC_list_lambda,
              nmcmc_tot = R))

}


validatePriors <- function(Priors) {

  if(is.null(Priors$mu_not) ||
     ncol(Priors$mu_not) != gremlinsEnv$num_parameters ||
     nrow(Priors$mu_not) != gremlinsEnv$nCovariates ||
     typeof(Priors$mu_not) != "double") {
    stop("mu_not should be a matrix of numerics values
         with columns equal to the number of parameters
         and rows equal to the number of covariates.")
  }

  if(is.null(Priors$Ainv) ||
     !is.matrix(Priors$Ainv) ||
     ncol(Priors$Ainv) != nrow(Priors$Ainv) ||
     ncol(Priors$Ainv) != gremlinsEnv$nCovariates ||
     typeof(Priors$Ainv) != "double") {
    stop("Ainv should be a square matrix of numeric values
         with dimensions equal to the number of covariates")
  }



  #TODO Finish validating the priors

}

drawPhi <- function(K, priors) {
  N_k <- double(length(priors))
  for(i in 1:length(priors)) {
    N_k[i] <- sum(K == i)
  }
  return(rdirichlet(N_k + priors))
}

#' @importFrom stats rgamma
rdirichlet <- function(alpha) {
  y <- rgamma(length(alpha), alpha)
  return(y/sum(y))
}
